lib/src/lib.rs

```diff
@@ -0,0 +1,389 @@
+#[derive(Debug)]
+pub enum TEEs {
+    TDX,
+    SGX,
+    SEVSNP,
+    SAMPLE,
+    NONE,
+}
+
+/// Supported Policy Engine types:
+/// - OPA: The Open Policy Egine.
+#[derive(Debug)]
+pub enum PolicyEngine {
+    OPA,
+    NONE,
+}
+
+/// Supported Policy Engine's dependency files:
+/// - Policy: OPA's policy(.rego) file.
+/// - Reference: OPA's reference data file.
+#[derive(Debug)]
+pub enum Files {
+    Policy,
+    Reference,
+    NONE,
+}
+
+#[derive(Debug)]
+pub struct Service {
+    pub locker: RwLock<u8>,
+    pub workdir: PathBuf,
+    pub attestation: core::Attestation,
+}
+
+impl Service {
+    /// Create a new attestation service's instance.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use attestation_service::Service;
+    /// use std::path::Path;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let workdir = Path::new("./");
+    ///     let user_id = "tester".to_string();
+    ///     let service = Service::new(&workdir, user_id).await;
+    ///     assert!(service.is_ok());
+    /// }
+    /// ```
+    pub async fn new(path: &Path, user_id: String) -> Result<Self> {
+        let dir = path.join(user_id);
+        Ok(Self {
+            locker: RwLock::new(0),
+            workdir: dir.clone(),
+            attestation: core::Attestation::new(dir.clone()).await?,
+        })
+    }
+
+    /// Attest the received Evidence by the attestation service instance.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use attestation_service::Service;
+    /// use std::path::Path;
+    /// use serde_json::json;
+    /// use sha2::{Sha384, Digest};
+    /// use base64;
+    ///
+    /// // sample TEE's evidence
+    /// fn evidence() -> String {
+    ///     let nonce = "the nonce".to_string();
+    ///     let public_key = "the pubic key".to_string();
+    ///     let mut hasher = Sha384::new();
+    ///     hasher.update(&nonce);
+    ///     hasher.update(&public_key);
+    ///     let hash = hasher.finalize();
+    ///     let quote = json!({
+    ///         "is_debuggable": false,
+    ///         "cpusvn": 1,
+    ///         "svn": 1,
+    ///         "report_data": base64::encode(hash)
+    ///     }).to_string();
+    ///     let ehd = json!({
+    ///         "nonce": nonce,
+    ///         "public_key": public_key
+    ///     }).to_string();
+    ///     json!({
+    ///         "tee": "sample".to_string(),
+    ///         "quote": quote,
+    ///         "ehd": ehd,
+    ///         "aad": "".to_string()
+    ///     }).to_string()
+    /// }
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let workdir = Path::new("./");
+    ///     let user_id = "tester".to_string();
+    ///     let service = Service::new(&workdir, user_id).await.unwrap();
+    ///     let res = service.attestation(&evidence()).await;
+    ///     assert!(res.is_ok());
+    /// }
+    /// ```
+    pub async fn attestation(&self, evidence: &String) -> Result<String> {
```

Comment:
Hi @dubek ,
I think It's better to let the invoker to convert this results to protobuf message and send it back to client. Because this `lib ` sub-project is a Crate. 
And I will add another project names `server` application which will depend on this Crate. It can response the protobuf format results.

Reply:
I agree that the lib (crate) should not return protobuf messages. But maybe it can returns a Rust struct with meaningful fields instead of String -- if it makes sense (that is, if there are several fixed fields with known meaning/use).

Reply:
Yes, this string is a JSON string and it is Deserialized by this structure:
```RUST
#[derive(Serialize, Deserialize, Debug)]
pub struct AttestationResults {
    pub tee: String,     // The TEE type, such as  "tdx", "sev-snp"
    pub result: String,   // The policy engine's evaluation reults.
    pub policy_info: String,   // It's also a JSON string, you can check the \src\core\verifier\sample\mod.rs for more info
    pub tcb: String,   // It's also a JSON string, please check the \src\core\verifier\sample\mod.rs for more info
}
```